File: ios/Domain/Models/BarcodeScanningResult+remap.swift
import Foundation

#if MLKIT_BARCODE_SCANNING
  #if canImport(MLKitVision)
    import MLKitVision

    extension BarcodeScanningResult {
      /// Remaps each barcode's bounding box and corner points back onto full-source-image
      /// coordinates, undoing the ROI crop offset and preprocessing scale.
      func remapped(using metadata: ImageMetadata) -> BarcodeScanningResult {
        return BarcodeScanningResult(
          barcodes: barcodes.map { $0.remapped(using: metadata) }
        )
      }
    }

    extension BarcodeData {
      fileprivate func remapped(using metadata: ImageMetadata) -> BarcodeData {
        return BarcodeData(
          bounds: bounds?.remapped(using: metadata),
          corners: corners?.map { $0.remapped(using: metadata) },
          format: format,
          formatName: formatName,
          valueType: valueType,
          valueTypeName: valueTypeName,
          rawValue: rawValue,
          displayValue: displayValue,
          rawBytes: rawBytes,
          isPotential: isPotential,
          value: value
        )
      }
    }
  #endif
#endif  // MLKIT_BARCODE_SCANNING

File: android/src/main/java/com/visioncameramlkit/domain/models/BarcodeScanningResult+remap.kt
package com.visioncameramlkit.domain.models

/**
 * Remaps every bounding box and corner in this result from processed-image coordinates
 * back to source image/frame coordinates, using [metadata]'s crop offset and scale
 * factor. See [BoundingBox.remap] and [Corner.remap] for the primitive.
 */
fun BarcodeScanningResult.remap(metadata: ImageMetadata): BarcodeScanningResult = copy(barcodes = barcodes.map { it.remap(metadata) })

private fun BarcodeData.remap(metadata: ImageMetadata): BarcodeData =
  copy(
    bounds = bounds?.remap(metadata),
    corners = corners?.map { it.remap(metadata) },
  )
